src/unix/tcp.c

```diff
@@ -28,86 +28,78 @@
 #include <errno.h>
 
 
-static int new_socket(uv_tcp_t* handle, int domain, unsigned long flags) {
-  struct sockaddr_storage saddr;
+static int maybe_bind_socket(int fd) {
+  union uv__sockaddr s;
   socklen_t slen;
+
+  slen = sizeof(s);
+  memset(&s, 0, sizeof(s));
+
+  if (getsockname(fd, &s.addr, &slen))
+    return UV__ERR(errno);
+
+  if (s.addr.sa_family == AF_INET)
+    if (s.in.sin_port != 0)
+      return 0;  /* Already bound to a port. */
+
+  if (s.addr.sa_family == AF_INET6)
+    if (s.in6.sin6_port != 0)
+      return 0;  /* Already bound to a port. */
+
+  /* Bind to an arbitrary port. */
+  if (bind(fd, &s.addr, slen))
+    return UV__ERR(errno);
```

Comment:
This code path doesn't appear to be exercised by our test suite at all...

The functionality was added, along with a test (spawn_tcp_server), in commit c5dd2d42 but that test doesn't get here at all. It calls uv_tcp_open(), which completely sidesteps this code path.

Something like this diff is needed to actually end up here:
```diff
diff --git a/src/unix/tcp.c b/src/unix/tcp.c
index cdc6ad8d..677b1b11 100644
--- a/src/unix/tcp.c
+++ b/src/unix/tcp.c
@@ -269,9 +269,13 @@ int uv_tcp_open(uv_tcp_t* handle, uv_os_sock_t sock) {
   if (err)
     return err;
 
-  return uv__stream_open((uv_stream_t*)handle,
+  err = uv__stream_open((uv_stream_t*)handle,
                          sock,
                          UV_HANDLE_READABLE | UV_HANDLE_WRITABLE);
+  if (err)
+    return err;
+
+  return maybe_bind_socket(sock);
 }
 
 
```
<sup>Note: for demonstration purposes only.</sup>